gax-java/gax/src/main/java/com/google/api/gax/logging/Slf4jUtils.java

```diff
@@ -0,0 +1,268 @@
+  }
+
+  interface LoggerFactoryProvider {
+    ILoggerFactory getLoggerFactory();
+  }
+
+  static class DefaultLoggerFactoryProvider implements LoggerFactoryProvider {
+    @Override
+    public ILoggerFactory getLoggerFactory() {
+      return LoggerFactory.getILoggerFactory();
+    }
+  }
+
+  // logging helper methods
+  static Map<String, Object> messageToMapWithGson(Message message)
+      throws InvalidProtocolBufferException {
+    String json = JsonFormat.printer().print(message);
+    return gson.fromJson(json, new TypeToken<Map<String, Object>>() {}.getType());
+  }
+
+  static void recordServiceRpcAndRequestHeaders(
+      String serviceName,
+      String rpcName,
+      String endpoint,
+      Map<String, String> requestHeaders,
+      LogData.Builder logDataBuilder,
+      LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isInfoEnabled()) {
+            addIfNotEmpty(logDataBuilder::serviceName, serviceName);
+            addIfNotEmpty(logDataBuilder::rpcName, rpcName);
+            addIfNotEmpty(logDataBuilder::httpUrl, endpoint);
+          }
+          if (logger.isDebugEnabled()) {
+            logDataBuilder.requestHeaders(requestHeaders);
+          }
+        });
+  }
+
+  private static void addIfNotEmpty(Consumer<String> setter, String value) {
+    if (value != null && !value.isEmpty()) {
+      setter.accept(value);
+    }
+  }
+
+  static void recordResponseHeaders(
+      Map<String, String> headers, LogData.Builder logDataBuilder, LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isDebugEnabled()) {
+            logDataBuilder.responseHeaders(headers);
+          }
+        });
+  }
+
+  static <RespT> void recordResponsePayload(
+      RespT message, LogData.Builder logDataBuilder, LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isDebugEnabled()) {
+            Map<String, Object> messageToMapWithGson =
+                Slf4jUtils.messageToMapWithGson((Message) message);
+
+            logDataBuilder.responsePayload(messageToMapWithGson);
+          }
+        });
+  }
+
+  static void logResponse(
+      String status, LogData.Builder logDataBuilder, LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isInfoEnabled()) {
+            logDataBuilder.responseStatus(status);
+          }
+          if (logger.isInfoEnabled() && !logger.isDebugEnabled()) {
+            Map<String, Object> responseData = logDataBuilder.build().toMapResponse();
+            Slf4jUtils.log(logger, Level.INFO, responseData, "Received Grpc response");
```

Comment:
IIUC, `logResponse` is used by both HttpJson and gRPC. I don't think we should have `Grpc` in the message here.

Reply:
thanks for catching!